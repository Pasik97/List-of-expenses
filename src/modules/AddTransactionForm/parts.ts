import styled from 'styled-components';
import media from 'utils/media';

export const AddTransactionFormWrapper = styled.div`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   margin: 24px 0;
`;

export const Row = styled.div`
   margin-bottom: 8px;

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 3fr 1fr;
   }
`;

export const InputWrapper = styled.div`
   margin: 0 0 8px 0;

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 1fr 1fr;
   }
`;

export const InputLabel = styled.p`
   white-space: nowrap;
   margin: 0;

   ${media.onlyMobile} {
      margin-bottom: 6px;
   }

   ${media.fromTablet} {
      flex-grow: 1;
   }
`;

export const Input = styled.input`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   border: none;
   outline: none;
   border-bottom: 1px solid #dddddd;

   &:focus {
      border-bottom: 1px solid #1a1a1a;
   }

   ${media.fromTablet} {
      flex-grow: 1;
   }
`;

export const AddButton = styled.button`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   background: none;
   border: 2px solid #dddddd;
   padding: 6px 24px;

   &:disabled {
      color: #dddddd;
   }

   &:hover, &active {
      border-color: #1a1a1a;
      cursor: pointer;
   }

   ${media.fromTablet} {
      height: fit-content;
      width: fit-content;
      justify-self: flex-end;
   }
`;

export const ValidateInputWrapper = styled.div`
   display: flex;
   flex-direction: column;

   ${media.fromTablet} {
      margin-left: 16px;
      flex-grow: 1;
   }
`;

export const Validation = styled.p`
   margin: 8px 0 0 0;
   color: red;
   font-size: 13px;
   line-height: 16px;
   min-height: 16px;
`;
