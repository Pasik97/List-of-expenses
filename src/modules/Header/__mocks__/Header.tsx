const mock = () => '(Header /)';

export default mock;
