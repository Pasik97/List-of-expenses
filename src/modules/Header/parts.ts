import styled from 'styled-components';

export const HeaderWrapper = styled.header`
   display: flex;
   align-items: baseline;
   justify-content: space-between;
   margin: 16px 0;
`;

export const Title = styled.h2`
   color: #1a1a1a;
   font-size: 28px;
   line-height: 32px;
   margin: 0 8px 0 0;
   font-weight: 700;
`;

export const ConversionRateWrapper = styled.div`
   display: flex;
   align-items: baseline;
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
`;

export const RateInput = styled.input`
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
   padding: 0;
   border: none;
   outline: none;
   border-bottom: 1px solid #dddddd;
   width: 60px;

   &:focus {
      border-bottom: 1px solid #1a1a1a;
   }
`;

export const ConversionLabel = styled.p`
   margin: 0;
   white-space: pre;
`;
