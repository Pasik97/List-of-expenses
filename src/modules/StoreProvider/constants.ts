import { ReactNode } from "react";
import Stores from "stores/Stores";

export interface StoreProviderProps {
   store: Stores;
   children: ReactNode;
}
