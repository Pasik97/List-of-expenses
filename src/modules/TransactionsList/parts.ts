import styled from 'styled-components';
import media from 'utils/media';

export const TransactionsListWrapper = styled.div``;

export const Cell = styled.div`
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
   border-right: 1px solid #1a1a1a;
   padding: 6px;
   display: flex;
   align-items: center;

   &:last-of-type {
      border-right: none;
   }
`;

export const HeaderCell = styled(Cell)`
   font-weight: 700;
   background-color: #cccccc;
   
   &:hover, &:active {
      cursor: pointer;
   }
`;

export const Row = styled.div`
   display: grid;
   grid-template-columns: 1fr 1fr 1fr 1fr;
   border: 1px solid #1a1a1a;
   border-top: none;

   &:first-of-type {
      border-top: 1px solid #1a1a1a;
   }

   &:nth-of-type(2n+3) {
      ${Cell} {
         background-color: #eeeeee;
      }
   }

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 2fr 1fr 1fr 1fr;
   }

   ${media.fromDesktop} {
      display: grid;
      grid-template-columns: 3fr 1fr 1fr 1fr;
   }
`;

export const Label = styled.p`
   margin: 0;
   overflow-wrap: anywhere;
`;

export const DeleteButton = styled.button`
   color: #4d4d4d;
   font-size: 16px;
   line-height: 21px;
   border: none;
   outline: none;
   background: none;

   &:hover, &:active {
      cursor: pointer;
      color: #1a1a1a;
   }
`;
