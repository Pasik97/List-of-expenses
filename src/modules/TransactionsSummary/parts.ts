import styled from 'styled-components';

export const Summary = styled.p`
   color: #1a1a1a;
   font-size: 18px; 
   line-height: 21px;
   margin: 16px 0 0 0;
`;
