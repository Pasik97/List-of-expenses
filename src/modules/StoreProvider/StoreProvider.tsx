import React, { createContext, } from 'react';
import { StoreProviderProps } from './constants';
import Stores from 'stores/Stores';

export const StoreProviderContext = createContext<Stores>({} as Stores);

const StoreProvider: React.FC<StoreProviderProps> = ({
   children,
   store,
}) => (
   <StoreProviderContext.Provider value={store}>
      {children}
   </StoreProviderContext.Provider>
);

export default StoreProvider;
