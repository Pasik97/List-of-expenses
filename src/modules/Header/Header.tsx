import React, { useContext, useState } from 'react';
import { observer } from 'mobx-react';
import * as P from './parts';
import { StoreProviderContext } from 'modules/StoreProvider/StoreProvider';
import { doubleNumberRegex } from 'utils/validators';

const Header: React.FC = observer(() => {
   const { transactions } = useContext(StoreProviderContext);

   const [rate, setRate] = useState<string>(`${transactions.conversionRate}`);

   const onRateChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const isRateValid = doubleNumberRegex.test(event.target.value);

      if (isRateValid || !event.target.value.length) {
         setRate(event.target.value);

         const newRate = Number(event.target.value) || 0;

         transactions.setConversionRate(newRate);
      }
   };

   return (
      <P.HeaderWrapper>
         <P.Title title="List of expenses">List of expenses</P.Title>
         <P.ConversionRateWrapper>
            <P.ConversionLabel>1EUR = </P.ConversionLabel>
            <P.RateInput name="rate" value={rate || ''} onChange={onRateChange} />
            <P.ConversionLabel>PLN</P.ConversionLabel>
         </P.ConversionRateWrapper>
      </P.HeaderWrapper>
   )
});


export default Header;