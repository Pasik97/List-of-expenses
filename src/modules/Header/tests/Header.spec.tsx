import { act, create } from "react-test-renderer";
import * as M from './mockData';
import Header from "../Header";
import { RateInput } from "../parts";
import StoreProvider from "modules/StoreProvider/StoreProvider";

describe('<Header />', () => {
   it('should render title and input to change conversion rate', () => {
      const component = create(
         <StoreProvider store={M.mockStore}>
            <Header />
         </StoreProvider>,
      );

      expect(component.toJSON()).toMatchSnapshot();
   });

   describe('onRateChange', () => {
      it('should set input value and call setConversionRate transactions class method if user entered valid number', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <Header />
            </StoreProvider>,
         );
         let rateInput = component.root.findByType(RateInput);

         act(() => {
            rateInput.props.onChange({ target: { value: '12.22' } });
         });

         rateInput = component.root.findByType(RateInput);
   
         expect(M.mockStore.transactions.setConversionRate).toHaveBeenCalled();
         expect(rateInput.props.value).toEqual('12.22');
      });

      it('should set input value and call setConversionRate transactions class method if clear input', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <Header />
            </StoreProvider>,
         );
         let rateInput = component.root.findByType(RateInput);

         act(() => {
            rateInput.props.onChange({ target: { value: '' } });
         });

         rateInput = component.root.findByType(RateInput);
   
         expect(M.mockStore.transactions.setConversionRate).toHaveBeenCalled();
         expect(rateInput.props.value).toEqual('');
      });

      it('should not set input value and not call setConversionRate transactions class method if user enter invalid value', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <Header />
            </StoreProvider>,
         );
         let rateInput = component.root.findByType(RateInput);

         expect(rateInput.props.value).toEqual(`${M.mockStore.transactions.conversionRate}`);

         act(() => {
            rateInput.props.onChange({ target: { value: 'ddd' } });
         });

         rateInput = component.root.findByType(RateInput);
   
         expect(M.mockStore.transactions.setConversionRate).not.toHaveBeenCalled();
         expect(rateInput.props.value).toEqual(`${M.mockStore.transactions.conversionRate}`);
      });
   });
});
