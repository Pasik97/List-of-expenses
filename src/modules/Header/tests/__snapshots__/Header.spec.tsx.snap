// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render title and input to change conversion rate 1`] = `
<header
  className="sc-bdvvtL jlYooz"
>
  <h2
    className="sc-gsDKAQ hOGUaI"
    title="List of expenses"
  >
    List of expenses
  </h2>
  <div
    className="sc-dkPtRN eyduwE"
  >
    <p
      className="sc-eCImPb ccvvFK"
    >
      1EUR = 
    </p>
    <input
      className="sc-hKwDye ghgza"
      name="rate"
      onChange={[Function]}
      value="12.11"
    />
    <p
      className="sc-eCImPb ccvvFK"
    >
      PLN
    </p>
  </div>
</header>
`;
